### DISCUSSION TITLE
Examples of ImagePipelineDelegate?

### DISCUSSION THREAD

**Discussion by treitter** [kean/Nuke#702] [Q&A]
Hi, I've got a custom `dataLoader` that makes composite images from local resources. I've got a `LazyImage` I'm using in SwiftUI that uses a custom pipeline that uses the custom `dataLoader`. Occasionally, I need to modify the behavior of the `dataLoader` (by changing some simple internal state) without tearing down the SwiftUI `View`. If I understand correctly, this is the use case for `ImagePipelineDelegate`.

Are there examples of how to implement a `ImagePipelineDelegate` (especially for this simple case)? I tried following the code itself but it's not quite clear to me which functions I need to implement and how. Just implementing `dataLoader(for:pipeline:)` appears to not be sufficient since none of my images render.

Thank you!

**Comment by treitter** [comment]
With more-careful reading of the code, it seems like the `ImagePipelineDelegate` is mostly limited to passive reaction to pipeline events (mainly for the purposes of logging and maybe debugging). I think it would be helpful to clarify that in the class documentation.

I managed to solve my problem by exposing my `DataLoader` so it can be modified between calls to `loadImage`. It seems to work well enough.

**Reply by kean** [reply]
`ImagePipelineDelegate` is a kitchen sink. 

The `dataLoader(for:pipeline:)` method is designed to be used for customizing which loader is used for a given request. The equivalent requests get coalesced, so it's not guaranteed to be called for each request.

>  I think it would be helpful to clarify that in the class documentation.

That's a good idea. I think it could also be split into separate protocols in the manner of the `URLSession` delegate because it does serve different purposes. The monitoring part should definitely be split off.

> I managed to solve my problem by exposing my DataLoader so it can be modified between calls to loadImage. It seems to work well enough.

That sounds good. I would appreciate it if you could share a bit more details about what the use case was.
